src/algorithms.jl

```diff
@@ -36,7 +36,7 @@ function DiffEqBase.remake(thing::Union{OrdinaryDiffEqAdaptiveImplicitAlgorithm{
                         OrdinaryDiffEqImplicitAlgorithm{CS,AD,FDT},
                         DAEAlgorithm{CS,AD,FDT}}; kwargs...) where {CS, AD, FDT}
   T = SciMLBase.remaker_of(thing)
-  T(; chunk_size=CS,autodiff=AD,SciMLBase.struct_as_namedtuple(thing)...,kwargs...)
+  T(; chunk_size=CS,autodiff=Val(AD),SciMLBase.struct_as_namedtuple(thing)...,kwargs...)
 end
```

Comment:
This constructor was type unstable, which was the motivation for the `_unwrap_val` change.

Be aware that while `AD` is a compile time constant here, passing it as a runtime value to `T` makes it a runtime value there.
It's possible for it to be inlined and const prop to take care of it, but that isn't guaranteed, and in fact was not happening, making this code needlessly type unstable.

Reply:
Awesome. For the tests of https://github.com/SciML/DifferentialEquations.jl/issues/786 it was showing inference working, but it's always better to make it explicit anyways.

Reply:
Yeah, this should be less brittle to different algorithms / Julia versions.
I was on Julia 1.6 with Rosenbrock23 (as part of a composite algorithm with Tsit5 IIRC).